File: app/src/main/java/forpdateam/ru/forpda/ui/fragments/mentions/MentionsFragment.java
package forpdateam.ru.forpda.ui.fragments.mentions;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.TabLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.arellomobile.mvp.presenter.InjectPresenter;
import com.arellomobile.mvp.presenter.ProvidePresenter;

import forpdateam.ru.forpda.App;
import forpdateam.ru.forpda.Di;
import forpdateam.ru.forpda.R;
import forpdateam.ru.forpda.api.mentions.models.MentionItem;
import forpdateam.ru.forpda.api.mentions.models.MentionsData;
import forpdateam.ru.forpda.client.ClientHelper;
import forpdateam.ru.forpda.presentation.mentions.MentionsPresenter;
import forpdateam.ru.forpda.presentation.mentions.MentionsView;
import forpdateam.ru.forpda.ui.fragments.RecyclerFragment;
import forpdateam.ru.forpda.ui.fragments.favorites.FavoritesHelper;
import forpdateam.ru.forpda.ui.views.ContentController;
import forpdateam.ru.forpda.ui.views.DynamicDialogMenu;
import forpdateam.ru.forpda.ui.views.FunnyContent;
import forpdateam.ru.forpda.ui.views.adapters.BaseAdapter;
import forpdateam.ru.forpda.ui.views.pagination.PaginationHelper;

/**
 * Created by radiationx on 21.01.17.
 */

public class MentionsFragment extends RecyclerFragment implements MentionsView {

    @InjectPresenter
    MentionsPresenter presenter;

    @ProvidePresenter
    MentionsPresenter provideMentionsPresenter() {
        return new MentionsPresenter(Di.get().mentionsRepository);
    }

    private DynamicDialogMenu<MentionsFragment, MentionItem> dialogMenu;
    private MentionsAdapter adapter;
    private PaginationHelper paginationHelper;

    public MentionsFragment() {
        configuration.setAlone(true);
        configuration.setDefaultTitle(App.get().getString(R.string.fragment_title_mentions));
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        paginationHelper = new PaginationHelper(getActivity());
        paginationHelper.addInToolbar(inflater, toolbarLayout, configuration.isFitSystemWindow());
        return view;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        viewsReady();

        dialogMenu = new DynamicDialogMenu<>();
        dialogMenu.addItem(getString(R.string.copy_link), (context, data) -> presenter.copyLink(data));
        dialogMenu.addItem(getString(R.string.add_to_favorites), (context, data) -> presenter.addToFavorites(data));

        adapter = new MentionsAdapter();

        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        recyclerView.setAdapter(adapter);

        adapter.setOnItemClickListener(adapterListener);
        refreshLayout.setOnRefreshListener(this::loadData);
        paginationHelper.setListener(paginationListener);

    }

    @Override
    public boolean loadData() {
        if (!super.loadData()) {
            return false;
        }
        presenter.getMentions(paginationHelper.getCurrentPage());
        return true;
    }

    @Override
    public void showMentions(MentionsData data) {
        if (data.getItems().isEmpty()) {
            if (!contentController.contains(ContentController.TAG_NO_DATA)) {
                FunnyContent funnyContent = new FunnyContent(getContext())
                        .setImage(R.drawable.ic_notifications)
                        .setTitle(R.string.funny_mentions_nodata_title)
                        .setDesc(R.string.funny_mentions_nodata_desc);
                contentController.addContent(funnyContent, ContentController.TAG_NO_DATA);
            }
            contentController.showContent(ContentController.TAG_NO_DATA);
        } else {
            contentController.hideContent(ContentController.TAG_NO_DATA);
        }

        adapter.addAll(data.getItems());
        paginationHelper.updatePagination(data.getPagination());
        setSubtitle(paginationHelper.getTitle());
        listScrollTop();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (paginationHelper != null)
            paginationHelper.destroy();
    }

    @Override
    public void showItemDialogMenu(MentionItem item) {
        dialogMenu.disallowAll();
        dialogMenu.allow(0);
        if (item.isTopic() && ClientHelper.getAuthState()) {
            dialogMenu.allow(1);
        }
        dialogMenu.show(getContext(), MentionsFragment.this, item);
    }

    @Override
    public void showAddFavoritesDialog(int id) {
        FavoritesHelper.addWithDialog(getContext(), aBoolean -> {
            Toast.makeText(getContext(), aBoolean ? getString(R.string.favorites_added) : getString(R.string.error_occurred), Toast.LENGTH_SHORT).show();
        }, id);
    }

    private PaginationHelper.PaginationListener paginationListener = new PaginationHelper.PaginationListener() {
        @Override
        public boolean onTabSelected(TabLayout.Tab tab) {
            return refreshLayout.isRefreshing();
        }

        @Override
        public void onSelectedPage(int pageNumber) {
            loadData();
        }
    };

    private BaseAdapter.OnItemClickListener<MentionItem> adapterListener = new BaseAdapter.OnItemClickListener<MentionItem>() {
        @Override
        public void onItemClick(MentionItem item) {
            presenter.onItemClick(item);
        }

        @Override
        public boolean onItemLongClick(MentionItem item) {
            presenter.onItemLongClick(item);
            return false;
        }
    };
}

File: app/src/main/java/forpdateam/ru/forpda/ui/fragments/mentions/MentionsFragment.kt
package forpdateam.ru.forpda.ui.fragments.mentions

import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.tabs.TabLayout
import forpdateam.ru.forpda.R
import forpdateam.ru.forpda.entity.remote.mentions.MentionItem
import forpdateam.ru.forpda.entity.remote.mentions.MentionsData
import forpdateam.ru.forpda.extensions.quillMoxyPresenter
import forpdateam.ru.forpda.model.AuthHolder
import forpdateam.ru.forpda.model.data.remote.api.favorites.FavoritesApi
import forpdateam.ru.forpda.presentation.mentions.MentionsPresenter
import forpdateam.ru.forpda.presentation.mentions.MentionsView
import forpdateam.ru.forpda.ui.fragments.RecyclerFragment
import forpdateam.ru.forpda.ui.fragments.favorites.FavoritesFragment
import forpdateam.ru.forpda.ui.views.ContentController
import forpdateam.ru.forpda.ui.views.DynamicDialogMenu
import forpdateam.ru.forpda.ui.views.FunnyContent
import forpdateam.ru.forpda.ui.views.adapters.BaseAdapter
import forpdateam.ru.forpda.ui.views.pagination.PaginationHelper
import ru.radiationx.quill.inject

/**
 * Created by radiationx on 21.01.17.
 */

class MentionsFragment : RecyclerFragment(), MentionsView {

    companion object {
        fun newInstance() = MentionsFragment()
    }

    private lateinit var dialogMenu: DynamicDialogMenu<MentionsFragment, MentionItem>
    private lateinit var adapter: MentionsAdapter
    private lateinit var paginationHelper: PaginationHelper
    private val authHolder by inject<AuthHolder>()

    private val paginationListener = object : PaginationHelper.PaginationListener {
        override fun onTabSelected(tab: TabLayout.Tab): Boolean {
            return refreshLayout.isRefreshing
        }

        override fun onSelectedPage(pageNumber: Int) {
            presenter.currentSt = pageNumber
            presenter.getMentions()
        }
    }

    private val adapterListener = object : BaseAdapter.OnItemClickListener<MentionItem> {
        override fun onItemClick(item: MentionItem) {
            presenter.onItemClick(item)
        }

        override fun onItemLongClick(item: MentionItem): Boolean {
            presenter.onItemLongClick(item)
            return false
        }
    }

    private val presenter by quillMoxyPresenter<MentionsPresenter>()

    init {
        configuration.defaultTitle = getString(R.string.fragment_title_mentions)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        paginationHelper = PaginationHelper(requireActivity())
        paginationHelper.addInToolbar(toolbarLayout, configuration.isFitSystemWindow)

        setScrollFlagsEnterAlways()


        dialogMenu = DynamicDialogMenu()
        dialogMenu.apply {
            addItem(getString(R.string.copy_link)) { _, data ->
                presenter.copyLink(data)
            }
            addItem(getString(R.string.add_to_favorites)) { _, data ->
                presenter.addToFavorites(data)
            }
        }

        adapter = MentionsAdapter()

        recyclerView.layoutManager = LinearLayoutManager(requireContext())
        recyclerView.adapter = adapter

        adapter.setOnItemClickListener(adapterListener)
        refreshLayout.setOnRefreshListener { presenter.getMentions() }
        paginationHelper.setListener(paginationListener)
    }

    override fun showMentions(data: MentionsData) {
        if (data.items.isEmpty()) {
            if (!contentController.contains(ContentController.TAG_NO_DATA)) {
                val funnyContent = FunnyContent(requireContext())
                    .setImage(R.drawable.ic_notifications)
                    .setTitle(R.string.funny_mentions_nodata_title)
                    .setDesc(R.string.funny_mentions_nodata_desc)
                contentController.addContent(funnyContent, ContentController.TAG_NO_DATA)
            }
            contentController.showContent(ContentController.TAG_NO_DATA)
        } else {
            contentController.hideContent(ContentController.TAG_NO_DATA)
        }

        adapter.addAll(data.items)
        paginationHelper.updatePagination(data.pagination)
        setSubtitle(paginationHelper.title)
        listScrollTop()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        paginationHelper.destroy()
    }

    override fun showItemDialogMenu(item: MentionItem) {
        dialogMenu.apply {
            disallowAll()
            allow(0)
            if (item.isTopic && authHolder.get().isAuth()) {
                allow(1)
            }
            show(requireContext(), this@MentionsFragment, item)
        }
    }

    override fun showAddFavoritesDialog(id: Int) {
        AlertDialog.Builder(requireContext())
            .setTitle(R.string.favorites_subscribe_email)
            .setItems(FavoritesFragment.getSubNames(requireContext())) { _, which ->
                presenter.addTopicToFavorite(id, FavoritesApi.SUB_TYPES[which])
            }
            .show()
    }

    override fun onAddToFavorite(result: Boolean) {
        Toast.makeText(
            requireContext(),
            if (result) getString(R.string.favorites_added) else getString(R.string.error_occurred),
            Toast.LENGTH_SHORT
        ).show()
    }
}
